### PR TITLE
Move Hub to TypeScript

### PR DESCRIPTION
**Pull request by matsjla**
Moves the Hub to TypeScript

### CODE REVIEW

File: static/hub.ts [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+    // Component Factories
+
+    private codeEditorFactory(container: GoldenLayout.Container, state: any): void {
+        // Ensure editors are closable: some older versions had 'isClosable' false.
+        // NB there doesn't seem to be a better way to do this than reach into the config and rely on the fact nothing
+        // has used it yet.
+        container.parent.config.isClosable = true;
+        const editor = new Editor(this, state, container);
+        this.editors.push(editor);
+    }
```
Comment (matsjla) [inline static/hub.ts:294]: This returns void, but in the layout.registerComponent, we are attempting to return it, possible bug?

File: static/hub.ts [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+            // valid, and componentName will be present on all registered Components
+            if (child.componentName === 'codeEditor') {
+                return this.findParentRowOrColumnOrStack(child);
+            } else {
+                if (child.isRow || child.isColumn || child.isStack) {
+                    const editor = this.findEditorInChildren(child);
+                    if (editor) return editor;
+                }
+            }
+            index++;
```
[the remark this answers is not in the document]
Comment (partouf) [reply static/hub.ts:239]: the elem can also be a Component, and then we do have componentName, so the type should probably be ContentItem | Component? https://github.com/golden-layout/golden-layout/blob/v1.5.9/src/js/items/Component.js#L13

File: static/hub.ts [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+    public addAtRoot(elem: GoldenLayout.ContentItem): void {
+        const rootFirstItem = this.layout.root.contentItems[0];
+        if (rootFirstItem) {
+            if (rootFirstItem.isRoot || rootFirstItem.isColumn) {
+                rootFirstItem.addChild(elem);
+            } else {
+                // @ts-expect-error -- GoldenLayout's types are messed up here?
+                const newRow: ContentItem = this.layout.createContentItem({
+                    type: 'row',
+                }, this.layout.root);
```
[the remark this answers is not in the document]
Comment (partouf) [reply static/hub.ts:266]: I'm not sure why the return type is void, it definitely returns an item or it throws

File: static/hub.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (matsjla) [reply static/hub.ts:239]: Doesn't look like there is a type for lm.items.Component though :/

File: static/hub.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (matsjla) [reply static/hub.ts:266]: Yeah I'll just mark it as their types are wrong

**Review by mattgodbolt — APPROVED** [review]
Looks good to me: thanks for helping slowly get us into modern times :)

**Pull request merged**